proposals/0030-markdoc.md

```diff
@@ -0,0 +1,327 @@
+- Start Date: 2023-02-28
+- Reference Issues: https://github.com/withastro/roadmap/issues/496
+- Implementation PR: https://github.com/withastro/astro/pull/6209
+
+# Summary
+
+This RFC introduces Markdoc support to Astro Content Collections.
+
+
+# Example
+
+Say you've authored a collection of blog posts using Markdoc. You can store these entries as a collection, identically to Markdown or MDX:
+
+```bash
+src/content/
+	blog/
+		post-1.mdoc
+		post-2.mdoc
+		post-3.mdoc
+...
+```
+
+Then, you can query entry frontmatter with the same `getCollection()` and `getEntryBySlug()` APIs:
+
+```astro
+---
+import { getCollection, getEntryBySlug } from 'astro:content';
+
+const blog = await getCollection('blog');
+const firstEntry = await getEntryBySlug('blog', 'post-1');
+---
+```
+
+You can also render Astro components from your Markdoc by configuring Markdoc [tags](https://markdoc.dev/docs/tags) and [nodes](https://markdoc.dev/docs/nodes). This example creates an `aside` tag for use in any Markdoc Content Collection entry:
+
+```js
+// astro.config.mjs
+import { defineConfig } from 'astro/config';
+import markdoc from '@astrojs/markdoc';
+
+// https://astro.build/config
+export default defineConfig({
+	integrations: [
+		markdoc({
+			tags: {
+				aside: {
+					render: 'Aside',
+          attributes: {
+            // Component props as attribute definitions
+            // See Markdoc's documentation on defining attributes
+            // https://markdoc.dev/docs/attributes#defining-attributes
+            type: { type: String },
+          }
+				},
+			},
+		}),
+	],
```

Comment:
could these be defined on the content level too? e.g.: `projects` have a different set of tags than `blog`, etc. I see we’re getting the `<Content/>` component from `entry.render()` which makes this viable. Perhaps later rather than sooner, as I believe this would come with its own type-safety challenges and require specialised tooling

Reply:
The idea's definitely come up! This would require some checks to see which config to apply, but it wouldn't be too difficult. Though we're not sure dividing configs per-collection is the cleanest way to divide these up. Did you hit a particular use case in your own projects you can share?

Reply:
Well that's the example really, projects haven't got `aside`, but have `carousel` etc, whereas blog doesn't. Not that it can't be made to work like that, but it'd be like comparing typescript unions with optionals: we _can_ make a single type full of optionals but if it's either or it often times makes sense to use unions.

Reply:
Fair enough! The thinking is you'll apply all of these tags globally, and only reach for the tags you want by-collection. Astro will only ship resources for a tag for content that uses it.  I see the optional vs. union parallel, but there's never a case where a tag is _required_ in a document, so I don't think it's quite the same?

Only trouble I could see is whether a given tag like `aside` _should act differently_ depending on the collection. So far, I haven't seen that become an issue. Again, we could make this happen, just want to find a case where globals are a blocker!